### PR TITLE
Fix 426: Storage classes for C style variadic functions

### PR DESCRIPTION
**Pull request by MoonlightSentinel**

### CODE REVIEW

File: src/dparse/parser.d [hunk 1]
```diff
@@ -5288,8 +5294,27 @@ class Parser
...
+                if (!currentIs(tok!"..."))
+                    return null;
+
+                // Reset to the beginning of the current parameters
+                index = startIdx;
+                allocator.rollback(cp);
+
+                node.hasVarargs = true;
+                mixin(parseNodeQ!(`node.varargsAttributes`, `VariadicArgumentsAttributes`));
+                expect(tok!"...");
```
Comment (Hackerpilot) [inline src/dparse/parser.d:5315]: I think that this should be `if (expect(tok!"...") is null) return null;`, as the parser functions should return null if there is an error. I know that not a lot of the other parse functions work this way right now, but that's a but that needs to be fixed some other time.

File: src/dparse/parser.d [hunk 2]
```diff
@@ -5301,6 +5326,54 @@ class Parser
...
+        ParameterAttribute[] buffer;
+        ownArray(buffer, attributes);
+        return buffer;
+    }
+
+    /**
+     * Parses an attribute of C-style variadic parameters.
+     *
+     * $(GRAMMAR $(RULEDEF variadicArgumentsAttribute):
+     *       $(RULE variadicArgumentsAttribute)+
```
Comment (Hackerpilot) [inline src/dparse/parser.d:5356]: This looks copied and pasted.

File: src/dparse/parser.d [hunk 3]
```diff
@@ -5301,6 +5326,54 @@ class Parser
...
+
+        if (!currentIsOneOf(tok!"const", tok!"immutable", tok!"shared", tok!"scope", tok!"return"))
+        {
+            error("`const`, `immutable`, `shared`, `scope`, `return` expected");
+            return null;
+        }
+
+        node.idType = current.type;
+        advance();
+        node.tokens = tokens[startIndex .. index];
```
Comment (Hackerpilot) [inline src/dparse/parser.d:5373]: Can't this array only ever have a single element in it? There's only a single call to `advance` here and there aren't any loops in this function.

Comment (WebFreak001) [reply src/dparse/parser.d:5373]: `.tokens` comes from the base AST class which is a range to the tokens some AST node is constructed from, so it's always a range of tokens (and this is the boilerplate similar to all other functions I assume)

Comment (MoonlightSentinel) [reply src/dparse/parser.d:5373]: Tried to follow the usual style, could also write it this way if preferred.

```suggestion
        node.tokens = tokens[index - 1 .. index];
```

File: src/dparse/parser.d (hunk 2 above)
Comment (MoonlightSentinel) [reply src/dparse/parser.d]: Fixed

File: src/dparse/parser.d (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dparse/parser.d]: Fixed

File: src/dparse/parser.d [hunk 4]
```diff
@@ -5301,6 +5326,58 @@ class Parser
...
+        ParameterAttribute[] buffer;
+        ownArray(buffer, attributes);
+        return buffer;
+    }
+
+    /**
+     * Parses an attribute of C-style variadic parameters.
+     *
+     * $(GRAMMAR $(RULEDEF variadicArgumentsAttribute):
+      *      $(LITERAL 'const')
```
Comment (Hackerpilot) [inline src/dparse/parser.d:5356]: Is there an extra space on this line?

Comment (MoonlightSentinel) [reply src/dparse/parser.d]: Not anymore

**Pull request merged**